server/src/test/java/org/opensearch/index/ShardIndexingPressureStoreTests.java

```diff
@@ -0,0 +1,190 @@
+        final ShardIndexingPressureTracker[] trackers = new ShardIndexingPressureTracker[NUM_THREADS];
+        for (int i = 0; i < NUM_THREADS; i++) {
+            int counter = i;
+            threads[i] = new Thread(() -> {
+                trackers[counter] = store.getShardIndexingPressureTracker(testShardId);
+            });
+            threads[i].start();
+        }
+
+        for (Thread t : threads) {
+            t.join();
+        }
+
+        for (int i = 0; i < NUM_THREADS; i++) {
+            assertEquals(tracker, trackers[i]);
+        }
+        assertEquals(1, store.getShardIndexingPressureHotStore().size());
+        assertEquals(1, store.getShardIndexingPressureColdStore().size());
+        assertEquals(tracker, store.getShardIndexingPressureHotStore().get(testShardId));
+        assertEquals(tracker, store.getShardIndexingPressureColdStore().get(testShardId));
+    }
+
+    public void testShardIndexingPressureStoreConcurrentGetAndCleanup() throws Exception {
+        ShardIndexingPressureTracker tracker = store.getShardIndexingPressureTracker(testShardId);
+        final int NUM_THREADS = scaledRandomIntBetween(100, 500);
+        final Thread[] threads = new Thread[NUM_THREADS];
+        for (int i = 0; i < NUM_THREADS; i++) {
+            threads[i] = new Thread(() -> {
+                ShardIndexingPressureTracker tracker1 = store.getShardIndexingPressureTracker(testShardId);
+                assertEquals(tracker, tracker1);
+                store.tryTrackerCleanupFromHotStore(tracker, () -> true);
+            });
+            threads[i].start();
+        }
+
+        for (Thread t : threads) {
+            t.join();
+        }
+
+        assertEquals(0, store.getShardIndexingPressureHotStore().size());
+        assertEquals(1, store.getShardIndexingPressureColdStore().size());
+        assertEquals(tracker, store.getShardIndexingPressureColdStore().get(testShardId));
+    }
+
+    public void testTrackerConcurrentEvictionFromColdStore() throws Exception {
+        int maxColdStoreSize =  ShardIndexingPressureStore.MAX_COLD_STORE_SIZE.get(settings);
+        final int NUM_THREADS = scaledRandomIntBetween(maxColdStoreSize * 2, maxColdStoreSize * 8);
+        final Thread[] threads = new Thread[NUM_THREADS];
+        for (int i = 0; i < NUM_THREADS; i++) {
+            int counter = i;
+            threads[i] = new Thread(() -> {
+                ShardId shardId = new ShardId("index", "uuid", counter);
+                ShardIndexingPressureTracker tracker = store.getShardIndexingPressureTracker(shardId);
+                store.tryTrackerCleanupFromHotStore(tracker, () -> true);
+            });
+            threads[i].start();
+        }
+
+        for (Thread t : threads) {
+            t.join();
+        }
+
+        assertEquals(0, store.getShardIndexingPressureHotStore().size());
+        assertTrue(store.getShardIndexingPressureColdStore().size() <= maxColdStoreSize + 1);
```

Comment:
Sorry but why do we need `maxColdStoreSize + 1`

Reply:
This is just for the boundary condition as cleanup happens when the next element is attempted insert post limit breach, since the condition is `>` and not `>=` in `shardIndexingPressureColdStore.size() > maxColdStoreSize`